Add Text Fragments article

tomayac opened the pull request:
Fixes #2501 

Changes proposed in this pull request:

- Adds the article

bokand commented on src/site/content/en/blog/text-fragments/index.md:
````diff
@@ -0,0 +1,414 @@
+#:~:text=[prefix-,]textStart[,textEnd][,-suffix]
+```
+
+Each of `prefix-`, `textStart`, `textEnd`, and `-suffix` will only match text within a single
````
it's implied but might be useful to make explicit that a full `textStart,textEnd` range _can_ span multiple blocks.

tomayac replied:
Added.

bokand commented on src/site/content/en/blog/text-fragments/index.md:
```diff
@@ -0,0 +1,414 @@
+
+Traditional element fragments can be combined with text fragments.
+It is perfectly fine to have both in the same URL, for example,
+to provide a meaningful fallback for browsers that do not support Text Fragments yet.
```
This doesn't really work for browsers that don't support text fragments, they won't know to strip out the `:~:` part so the browser won't find any fragment.

The more useful case here is if text on the page changes so the text fragment no longer matches, a conforming browser will then fall back to the element id which might be the nearest section header, for example.

tomayac replied:
Changed to what you suggested.

bokand commented on src/site/content/en/blog/text-fragments/index.md:
```diff
@@ -0,0 +1,414 @@
+to create an extension for browsers that do not support Text Fragments natively
+where the functionality is implemented in JavaScript.
+
+### Security
```
It should be noted somewhere that the text fragment is only apply to the main frame - they won't search for text inside an iframe and an iframe navigation won't invoke a text fragment.

For background, this was restricted to reduce the attack surface - we don't have any known specific attacks and this is something we could consider loosening in the future but we thought the reduced attack surface would simplify things initially and we didn't have a strong use case.

tomayac replied:
Added.

bokand commented on src/site/content/en/blog/text-fragments/index.md:
We're experimenting with this using an origin trial in M83 (applying this policy but via the origin trial, rather than document policy which is as-yet unshipped). This might be a good place to mention the trial @nickburris

nickburris replied:
Agreed, let's link to the origin trial page here. I'm just waiting for the origin trials team to start the trial, I should have a link for you this week.

tomayac replied:
Added.

grantjwang commented:
I'm a github n00b, so apologies if my comments here aren't as github-friendly as they could be.

LGTM.

My main critique is that the screenshots that show highlighted text aren't as *obviously* screenshots as they could be; at times, they visually look like the rest of the article. A few ideas here: perhaps making the border of the screenshot more significant, making the screenshot smaller (so that it doesn't take the same width as the text of the article), or having the screenshot include browser chrome?

Also, I think your extension should be a bit more prominent in the article :).

tomayac commented:
> My main critique is that the screenshots that show highlighted text aren't as obviously screenshots as they could be; at times, they visually look like the rest of the article. A few ideas here: perhaps making the border of the screenshot more significant, making the screenshot smaller (so that it doesn't take the same width as the text of the article), or having the screenshot include browser chrome?

I have downsized them all, so they stand clearly out as screenshots.

> Also, I think your extension should be a bit more prominent in the article :).

Let's see :-)

jpmedley commented on src/site/content/en/blog/text-fragments/index.md:
```diff
@@ -0,0 +1,422 @@
+User agent instructions are also called *directives*.
+In the concrete case, `text=` is therefore called a *text directive*.
+
+### Feature detectability
```
```suggestion
### Feature detection
```

jpmedley commented on src/site/content/en/blog/text-fragments/index.md:
```diff
@@ -0,0 +1,422 @@
+
+### Feature detectability
+
+Support for Text Fragments is feature-detectable
+by checking for the existence of the `fragmentDirective` in the `Location` interface.
+Unlike the other properties like `hash` that I have mentioned above,
+`fragmentDirective` does not return any value and is read-only.
```
```suggestion
To detect support, test for the read only `fragmentDirective` property on `window.location`.
```

jpmedley commented on src/site/content/en/blog/text-fragments/index.md:
```diff
@@ -0,0 +1,422 @@
+It is important that implementations of the Text Fragments specification
+do not leak whether a text fragment was found on a page or not.
+While element fragments are fully under the control of the original page author
+(remember how there was no way to link to the section
+I wanted to link to before in my running example),
```
What?

tomayac replied:
There was no element identifier `id="something"` in the blog post, so I couldn't link.

jpmedley replied:
Then that's how you should say it. It's easier to understand. Also I'd put this after the current sentence.

tomayac replied:
I have rephrased.

jpmedley commented on src/site/content/en/blog/text-fragments/index.md:
```diff
@@ -0,0 +1,422 @@
+text fragments can be created by anyone.
+
+Imagine I ran an evil ad network `evil-ads.example.com`.
+Further imagine that in one of my ad iframes I opened a hidden cross-origin iframe
```
```suggestion
Further imagine that in one of my ad iframes I opened a hidden cross-origin iframe link
```

jpmedley commented on src/site/content/en/blog/text-fragments/index.md:
```suggestion
the Chrome team decided to implement scroll on navigation,
```

jpmedley commented on src/site/content/en/blog/text-fragments/index.md:
I do't understand why scroll navigation makes text fragments safer.

tomayac replied:
The opposite. We believe the risk scroll navigation adds is manageable.

jpmedley replied:
Your wording doesn't convey that.

tomayac replied:
I have rephrased.

tomayac commented on src/site/content/en/blog/text-fragments/index.md:
```diff
@@ -0,0 +1,422 @@
+---
+title: |
+  Link to where no human has linked to before: Text Fragments
```
The full frontal nerdity it is :-)

tomayac commented on src/site/content/en/blog/text-fragments/index.md:
```diff
@@ -0,0 +1,422 @@
+
+In its simplest form, the syntax of Text Fragments is as follows:
+The hash symbol `#` followed by `:~:text=` and finally `textStart`, which represents the
+[percent-encoded](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent)
```
In this case I disagree. The term "percent-encoded" appears in the RFC https://tools.ietf.org/html/rfc3986.

tomayac commented on src/site/content/en/blog/text-fragments/index.md:
```diff
@@ -0,0 +1,422 @@
+
+Taking up the example from above where I wanted to place a deep link to the
+*ECMAScript Modules in Web Workers* heading, the URL in this case would be
+<a href="https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html#:~:text=ECMAScript%20Modules%20in%20Web%20Workers"><code>https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html<strong>#:~:text=ECMAScript%20Modules%20in%20Web%20Workers</strong></code></a>
```
Yeah, I wish I could inline-style it. @kaycebasques, is it OK to add inline styles like `<em style="color: red;">`?

jpmedley replied:
I would add "See it for yourself" to the next paragraph and move the link to it. That would make the bolding easier to see and the intent of the link clearer. Do the same at line 58.

tomayac replied:
I have solved this now by using `<mark class="highlight-line highlight-line-active">`.

tomayac commented on src/site/content/en/blog/text-fragments/index.md:
```diff
@@ -0,0 +1,422 @@
+Like the other parameters, they, too, need to be percent-encoded.
+<a href="https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html#:~:text=the-,text,-parameter"><code>https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html<strong>#:~:text=the-,text,-parameter</strong></code></a>.
+To allow the parser to clearly identify the `prefix-` and the `-suffix`,
+they need to be separated from the `textStart` and the optional `textEnd` with a dash&nbsp;`-`.
```
`prefix` and `suffix` can contain multiple words, yes.

jpmedley replied:
You should mention that somewhere.

tomayac replied:
Added.

tomayac commented on src/site/content/en/blog/text-fragments/index.md:
```diff
@@ -0,0 +1,422 @@
+
+So far I have only quickly glanced over one important addition without actually explaining it:
+the fragment directive `:~:`.
+To avoid compatibility issues with the usage of existing URL element fragments as shown above,
```
The `#HTML1` part.

tomayac commented on src/site/content/en/blog/text-fragments/index.md:
```diff
@@ -0,0 +1,422 @@
+  <figcaption class="w-figcaption">Text Fragments announcement blog post excerpt.</figcaption>
+</figure>
+
+If I wanted to link to the word `text`
+that is written in a green code font at the bottom of the screenshot above,
+I would set `textStart` to `text`.
+Since the word `text` is, well, only one word, there cannot be a `textEnd`,
+but even just the blog post excerpt alone contains four times the word `text`.
+What now?
+The URL
+<a href="https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html#:~:text=text"><code>https://blog.chromium.org/2019/12/chrome-80-content-indexing-es-modules.html<strong>#:~:text=text</strong></code></a>
+matches at the first occurrence of the word `Text`.
```
The important part here is that "text" appears four times in this section alone, but I want to refer explicitly to the green code font one, not the other three.

jpmedley replied:
The way to handle this is to start the section with something like "Notice the green text in the image above." Then explain what you want readers to know about it.

tomayac replied:
Rephrased now.

kaycebasques commented:
@jpmedley sounds like this is waiting on a second review from you

jpmedley commented on src/site/content/en/blog/text-fragments/index.md:
```diff
@@ -0,0 +1,423 @@
+<figure class="w-figure">
+  <img src="correct-text.png" alt="" class="w-screenshot" width="400">
+  <figcaption class="w-figcaption">Text Fragment matching at the correct occurrence of <code>Text</code>.</figcaption>
+</figure>
```
@robdodson Is there a way to make screenshots take the whole width. What he's illustrating is difficult to see. https://deploy-preview-2503--web-dev-staging.netlify.app/text-fragments/

robdodson replied:
you can use `width="100%"` instead of `width="400"`

tomayac replied:
I have made these three screenshots wider, since their text is important.

tomayac commented:
Thanks for the additional review round, @jpmedley, I think this is now in a good state. @kaycebasques, if you want to have a look, please go for it. FYI, we're waiting for some changes to land in Chrome that @bokand is working on and for the extension mentioned in the article to be launched following the "by Google" process.

kaycebasques commented on src/site/content/en/blog/text-fragments/index.md:
```diff
@@ -0,0 +1,435 @@
+
+<figure class="w-figure">
+  <img src="id-missing.png" alt="" class="w-screenshot" width="400">
+  <figcaption class="w-figcaption">Dev Tools showing a heading without an <code>`id`</code>.</figcaption>
```
```suggestion
  <figcaption class="w-figcaption">Dev Tools showing a heading without an <code>id</code>.</figcaption>
```

kaycebasques commented on src/site/content/en/blog/text-fragments/index.md:
```diff
@@ -0,0 +1,435 @@
+The [Text Fragments](https://wicg.github.io/ScrollToTextFragment/) proposal adds support
+for specifying a text snippet in the URL hash.
+When navigating to a URL with such a text fragment, the user agent can emphasize
+and/or bring it to the user's attention.
+
+### `textStart`
```
```suggestion
and/or bring it to the user's attention.

### Browser compatibility

Text Fragments are supported in version 80 and beyond of Chromium-based browsers.
At the time of writing, Safari and Firefox have not publicly signaled an intent to implement
the feature. See [Related links](#related-links) for links to the Safari and Firefox discussions.

### `textStart`
```

kaycebasques commented on src/site/content/en/blog/text-fragments/index.md:
```diff
@@ -0,0 +1,435 @@
+<figure class="w-figure">
+  <img src="id-html1.png" alt="" class="w-screenshot" width="400">
+  <figcaption class="w-figcaption">Dev Tools showing the <code>id</code> of an element.</figcaption>
+</figure>
```
This image is rather small on account of `width="400"`

![image](https://user-images.githubusercontent.com/4713486/81128962-9bca1580-8ef7-11ea-8d43-98111abad6c3.png)

tomayac replied:
Now 600 :-)

kaycebasques commented on src/site/content/en/blog/text-fragments/index.md:
```diff
@@ -0,0 +1,435 @@
+This is the problem that Text Fragments solve 🎉.
+
+<figure class="w-figure">
+  <img src="id-missing.png" alt="" class="w-screenshot" width="400">
```
This image is also small on account of `width="400"`

![image](https://user-images.githubusercontent.com/4713486/81129002-b8fee400-8ef7-11ea-9a9e-a8e58037c24c.png)

kaycebasques commented on src/site/content/en/blog/text-fragments/index.md:
```diff
@@ -0,0 +1,435 @@
+highlighted and scrolls into view:
+
+<figure class="w-figure">
+  <img src="syntax-simple.png" alt="" class="w-screenshot" width="400">
```
Also small but perhaps this one was intentional

![image](https://user-images.githubusercontent.com/4713486/81129043-e186de00-8ef7-11ea-961e-b83ac16e3d77.png)

tomayac replied:
Intentional indeed. Just the heading is important.

tomayac commented on src/site/content/en/blog/text-fragments/index.md:
```diff
@@ -0,0 +1,441 @@
+
+### Browser compatibility
+
+Text Fragments are supported in version 80 and beyond of Chromium-based browsers.
```
```suggestion
The Text Fragments feature is supported in version 80 and beyond of Chromium-based browsers.
```

tomayac commented on src/site/content/en/blog/text-fragments/index.md:
```diff
@@ -0,0 +1,441 @@
+  and/or bring it to the user's attention.
+authors:
+  - thomassteiner
+date: 2020-04-03
```
```suggestion
date: 2020-05-06
```

tomayac commented on src/site/content/en/blog/text-fragments/index.md:
```diff
@@ -0,0 +1,441 @@
+has an `id` attribute with the value `HTML1`.
+
+<figure class="w-figure">
+  <img src="id-html1.png" alt="" class="w-screenshot" width="400">
```
```suggestion
  <img src="id-html1.png" alt="" class="w-screenshot" width="600">
```

tomayac commented on src/site/content/en/blog/text-fragments/index.md:
```diff
@@ -0,0 +1,441 @@
+This is the problem that Text Fragments solve 🎉.
+
+<figure class="w-figure">
+  <img src="id-missing.png" alt="" class="w-screenshot" width="400">
```
```suggestion
  <img src="id-missing.png" alt="" class="w-screenshot" width="600">
```

Now 600 :-)

kaycebasques commented:
@tomayac you should now have the ability to merge the PR when it's ready

bokand commented:
Just some last minute feedback.

Re: feature-detection; it seems `window.location` might not be the best place for this. It's been suggested we move it to `document.fragmentDirective`. See crbug.com/1057795. Should we perhaps mention this might change in a future release?

Also, it would be worth clarifying that the feature detection is mainly intended in cases where links are dynamically generated (e.g. search engine) to avoid serving such links to browsers that don't support it. There's some small risk of the page being negatively affected by a text fragment if it doesn't support it due to seeing an unexpected fragment (though most should load fine).

Should we also mention that these links currently don't work when served across a client-side redirect (HTTP redirects, e.g. bit.ly,  do work) and that some common services use them (Twitter, Hangouts Chat)? I'm not sure we should block on getting that fixed as there's still some uncertainty on how to best address this (it does look like changing the Google redirector will work at least for Google services).

jpmedley commented on src/site/content/en/blog/text-fragments/index.md:
I hate to throw a monkey wrench into this at the 11th hour. It looks like the opt-out code laded back in mid-April. No one told us or even bothered to [close the ticket until last Thursday](https://bugs.chromium.org/p/chromium/issues/detail?id=1047900). (Here's a link showing that this [merged to 83](https://storage.googleapis.com/chromium-find-releases-static/942.html#942995187c607e433af35e114a325c4ca2576d37) some time ago.)

cc: @kaycebasques, @petele

bokand replied:
Sorry about the confusion (I've reopened that bug). The thing that landed and was merged back to M83 was the origin trial - the DocumentPolicy remains unshiped. So I think this text is still correct as-is.

tomayac commented:
All changes were made in https://github.com/GoogleChrome/web.dev/pull/2503/commits/b9830cfcb87f9cc0cfc33a22330de7efa84d36f3.

> Re: feature-detection; it seems `window.location` might not be the best place for this. It's been suggested we move it to `document.fragmentDirective`. See crbug.com/1057795. Should we perhaps mention this might change in a future release?

Added. We have a nice caution aside that works for these kind of remarks.

> Also, it would be worth clarifying that the feature detection is mainly intended in cases where links are dynamically generated (e.g. search engine) to avoid serving such links to browsers that don't support it. There's some small risk of the page being negatively affected by a text fragment if it doesn't support it due to seeing an unexpected fragment (though most should load fine).

Added. This makes sense.

> Should we also mention that these links currently don't work when served across a client-side redirect (HTTP redirects, e.g. bit.ly, do work) and that some common services use them (Twitter, Hangouts Chat)? I'm not sure we should block on getting that fixed as there's still some uncertainty on how to best address this (it does look like changing the Google redirector will work at least for Google services).

Added. There is a browser compatibility section where it fits in nicely.

kaycebasques commented:
@tomayac do you want to merge this?

tomayac commented:
> @tomayac do you want to merge this?

We're still waiting for the extension to be released and be given the official "by Google" seal of approval. The COVID-19 situation slows down the review process… :-( For now, please just keep it ready to be merged (everyone as LGTM'ed) and don't remove the "do not merge" label. Thanks!
